### ISSUE TITLE
Debugging 'iptables: No chain/target/match by that name' and no external connectivity

### ISSUE THREAD

**Issue by jforman** [cloudnativelabs/kube-router#775]
Hi All,

I've got a kubeadm-created cluster on 1.15.2 where I'm trying to replace kube-proxy with kube-router. I'm at the point of bringing up kube-router with the full feature set turned on (routing, firewall, service proxy) but am seeing the following in my kube-router logs on the machine itself

```
2019-08-30T01:08:02.876189174+00:00 stderr F I0830 01:08:02.875956       1 kube-router.go:207] Running /usr/local/bin/kube-router version v0.3.2, built on 2019-07-25T11:50:15+0530, go1.10.8
2019-08-30T01:08:03.562407669+00:00 stderr F I0830 01:08:03.561704       1 network_policy_controller.go:143] Starting network policy controller
2019-08-30T01:08:03.614233021+00:00 stderr F I0830 01:08:03.614113       1 network_routes_controller.go:983] Could not find annotation `kube-router.io/bgp-local-addresses` on node object so BGP will listen on node IP: 10.10.2.138 address.
2019-08-30T01:08:03.624289298+00:00 stderr F time="2019-08-30T01:08:03Z" level=warning msg="Running modprobe ip_vs failed with message: `modprobe: module ip_vs not found in modules.dep`, error: exit status 1"
2019-08-30T01:08:03.629174996+00:00 stderr F time="2019-08-30T01:08:03Z" level=error msg="Could not get ipvs family information from the kernel. It is possible that ipvs is not enabled in your kernel. Native loadbalancing will not work until this is fixed."
2019-08-30T01:08:03.696131181+00:00 stderr F I0830 01:08:03.695994       1 network_services_controller.go:284] Starting network services controller
2019-08-30T01:08:04.095556339+00:00 stderr F E0830 01:08:04.095461       1 network_routes_controller.go:175] Error enabling Pod egress: Failed to add iptables rule to masquerade outbound traffic from pods: running [/sbin/iptables -t nat -A POSTROUTING -m set --match-set kube-router-pod-subnets src -m set ! --match-set kube-router-pod-subnets dst -m set ! --match-set kube-router-node-ips dst -j MASQUERADE --wait]: exit status 1: iptables: No chain/target/match by that name.
2019-08-30T01:08:04.095556339+00:00 stderr F External connectivity will not work.
```

What am I doing wrong in my iptables rules that this won't work?

For context sake, output from iptables:
```
root@prod-controller0:~# iptables -L -v -n
Chain INPUT (policy ACCEPT 712K packets, 107M bytes)
 pkts bytes target     prot opt in     out     source               destination         
 758K  145M KUBE-FIREWALL  all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
 1012 62783 ACCEPT     all  --  *      ens2    0.0.0.0/0            0.0.0.0/0            /* allow outbound node port traffic on node interface with which node ip is associated */
    0     0 ACCEPT     all  --  *      kube-bridge  0.0.0.0/0            0.0.0.0/0            /* allow inbound traffic to pods */
    0     0 ACCEPT     all  --  kube-bridge *       0.0.0.0/0            0.0.0.0/0            /* allow outbound traffic from pods */

Chain OUTPUT (policy ACCEPT 709K packets, 115M bytes)
 pkts bytes target     prot opt in     out     source               destination         
 734K  121M KUBE-FIREWALL  all  --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain KUBE-FIREWALL (2 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000


root@prod-controller0:~# iptables -L -v -n -t nat
Chain PREROUTING (policy ACCEPT 6 packets, 360 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain INPUT (policy ACCEPT 6 packets, 360 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 6 packets, 360 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain POSTROUTING (policy ACCEPT 6 packets, 360 bytes)
 pkts bytes target     prot opt in     out     source               destination         
 4868  296K KUBE-POSTROUTING  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes postrouting rules */

Chain KUBE-MARK-DROP (0 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK or 0x8000

Chain KUBE-MARK-MASQ (0 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK or 0x4000

Chain KUBE-POSTROUTING (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 MASQUERADE  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes service traffic requiring SNAT */ mark match 0x4000/0x4000
```

Any help is greatly appreciated.

**Comment by aauren** [comment]
That's pretty odd... I can't think what would cause that since POSTROUTING is a built in chain and from your `-L` output it obviously exists. You get different error messages if one of the `ipset`s or the `MASQUERADE` module doesn't exist.

I would recommend clearing out your iptables and rebooting and seeing if its still a problem. If it's still a problem after that maybe try executing the same command that same `iptables` command that `kube-router` is executing from the CLI. If your error message is still vague, maybe try bisecting it a bit by changing it around and see if you can get it to succeed.

**Comment by jforman** [comment]
@aauren I will note I cleared out the iptables manually 'iptables -F' and iptables -X for both regular and nat tables, and didnt use kube-proxy --cleanup. Not sure if that matters.

As for trying to narrow down the failing iptables rule, I tried them each individually after a reboot. All failed (unless you meant something else)

root@prod-controller0:~# /sbin/iptables -t nat -A POSTROUTING -m set --match-set kube-router-pod-subnets src -j MASQUERADE --wait
iptables: No chain/target/match by that name.

root@prod-controller0:~# /sbin/iptables -t nat -A POSTROUTING -m set ! --match-set kube-router-pod-subnets dst -j MASQUERADE --wait
iptables: No chain/target/match by that name.

root@prod-controller0:~# /sbin/iptables -t nat -A POSTROUTING -m set ! --match-set kube-router-node-ips dst -j MASQUERADE --wait
iptables: No chain/target/match by that name.

**Comment by jforman** [comment]
I wonder, is there a kernel module I'm missing for the 'set' or 'match' on the NAT table?

**Comment by aauren** [comment]
Yes, there is an iptables module for ip sets. If you have your config handy or if you have it available via `/proc` you can look for it.
```
zgrep IP_SET /proc/config.gz 
CONFIG_IP_SET=m
...
```

You could also try `modprobe ip_set` and see if it works. Or you could try using a different match option on the above iptables rules and see if everything works other than the ip set match logic. i.e. can you add a rule that targets MASQUERADE on the POSTROUTING chain of the nat table but doesn't use ip set logic.

**Comment by jforman** [comment]
> Yes, there is an iptables module for ip sets. If you have your config handy or if you have it available via `/proc` you can look for it.
> 
> ```
> zgrep IP_SET /proc/config.gz 
> CONFIG_IP_SET=m
> ...
> ```

root@prod-controller0:/boot# zgrep IP_SET config-4.15.0-1042-kvm
CONFIG_IP_SET=y
CONFIG_IP_SET_MAX=256
CONFIG_IP_SET_BITMAP_IP=m
......

root@prod-controller0:/boot# lsmod | grep ip_set
ip_set_hash_ip         24576  1
ip_set_hash_net        28672  1


> 
> You could also try `modprobe ip_set` and see if it works. Or you could try using a different match option on the above iptables rules and see if everything works other than the ip set match logic. i.e. can you add a rule that targets MASQUERADE on the POSTROUTING chain of the nat table but doesn't use ip set logic.

The more I look at this, the more I think it has to do with kernel modules, or perhaps some weird iptables syntax. 

Would you mind sharing your `lsmod` output? 

The lists in ipset are correct
```
root@prod-controller2:~# ipset list | grep Name
Name: kube-router-pod-subnets
Name: kube-router-node-ips
```

Trying a different rule destined for MASQUERADE works:

# /sbin/iptables -t nat -D POSTROUTING -p tcp --dport 1234 -j MASQUERADE

So, something awry with the 'set' iptables module?

**Comment by aauren** [comment]
Sorry this post is so late... Here are the iptables/net related modules that we have loaded:
```
veth                    9147  0
xt_mark                 1981  4
dummy                   9245  0
xt_nat                  4023  231
xt_set                 15260  6
ip_set_hash_ipport     34642  1
ipt_MASQUERADE          2171  3
nf_nat_masquerade_ipv4     3349  1 ipt_MASQUERADE
xt_comment              1571  11
ip_set_hash_ip         33258  3
ip_set_hash_net        37988  1
ip_set                 44174  4 ip_set_hash_ip,xt_set,ip_set_hash_net,ip_set_hash_ipport
xt_addrtype             4493  2
iptable_security        2765  0
iptable_nat             3423  1
nf_nat_ipv4             8253  1 iptable_nat
nf_nat                 28529  3 nf_nat_masquerade_ipv4,nf_nat_ipv4,xt_nat
iptable_raw             2676  0
iptable_mangle          3195  0
```

I definitely think it has something to do with your kernel modules.

**Comment by jforman** [comment]
So, I think I've found it. `xt_set` is a module on your system, and not a module nor compiled in on mine (Ubuntu 18.04.3 LTS cloud images). 

```
root@prod-controller1:/boot# grep XT_SET config-4.15.0-1043-kvm 
# CONFIG_NETFILTER_XT_SET is not set
```

Which should have been pretty obvious for me in poking around. I'm going to confirm by attempting to build a custom kernel, but in the meantime, I've filed https://bugs.launchpad.net/ubuntu/+source/linux-kvm/+bug/1843051 to get this included in Ubuntu. Hope this helps others who found this Issue. I will keep it updated with what I find.

**Comment by jforman** [comment]
It looks like `xt_set` was the magic kernel module. It also just so happens that that module is included in the 5.0.x kernels for Ubuntu 18.04 LTS.

Upgrading to `5.0.0-27-generic` was enough to get me past the above iptables issues, and on to something else (I cannot contact the core-dns service from within a pod on a different node than one of the pods for the service, but that is another issue.)

Hopefully this proves helpful for people who come across this issue!
